Added Bfloat16 tensor for cpu with very limited support

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21860 [WIP] Added Bfloat16 tensor for cpu with very limited support**
* #21523 [WIP] Enabled BFloat16 storage
* #21522 [WIP] Intoducing bfloat16 type

Differential Revision: [D15856091](https://our.internmc.facebook.com/intern/diff/D15856091)

------------
Enabled bfloat16 on CPU for all methods in declaration.cwrap which are supported by Half.
Tested via unit tests

izdeby committed:
Added Bfloat16 tensor for cpu with very limited support

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "[WIP] Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

izdeby committed:
Update on "Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

gchanan commented on test/test_torch.py:
```diff
@@ -2855,6 +2859,7 @@ def test_dtypes(self):
         all_dtypes = torch.testing.get_all_dtypes()
         do_test_dtypes(self, all_dtypes, torch.strided, torch.device('cpu'))
         if torch.cuda.is_available():
+            all_dtypes.remove(torch.bfloat16)  # Remove once _th_zero_ is enabled on cuda for bfloat16
```
as with the unfold test below, it's better to run the actual test and assert it raises an exception so you force the test to get updated.  As it is, we'll probably forget to actually test this.

izdeby replied:
Fixed for all the tests in this file but cant update this test due to a weird behavior in common_utils.py. This test will be updated in the next PR in the stack and CUDA path will work.

ezyang replied:
> cant update this test due to a weird behavior in common_utils.py

OOC, what kind of behavior?

izdeby replied:
In common_utils.py, in do_test_dtypes i would add something like: 
`if dtype == torch.bfloat16 and device == 'cuda:0':` but it will be ignored even when
dtype value is torch.bfloat16 and device is "cuda:0". The problem was with comparing the device variable value to 'cuda:0' - it would never pass.

Same check would work in test_torch.py with no problems.

ezyang replied:
I was not aware that `device` comparison with a string was a supported operation :) What I would expect to work is if you did `device == torch.device('cuda:0')`

gchanan commented on test/test_torch.py:
```diff
@@ -3333,6 +3338,9 @@ def test_bool_tensor_value_change(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if dt == torch.bfloat16:
+                    continue  # fix once random is implemented for bfloat16 on cuda
```
here too.

gchanan commented on test/test_torch.py:
```diff
@@ -3344,6 +3352,8 @@ def test_copy_all_dtypes_and_devices(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
+                    continue  # Remove once cuda available for bfloat16
```
here too.

gchanan commented on test/test_torch.py:
```diff
@@ -3358,13 +3368,17 @@ def test_resize_all_dtypes_and_devices(self):
         shape = (2, 2)
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -3358,13 +3368,17 @@ def test_resize_all_dtypes_and_devices(self):
     def test_resize_as_all_dtypes_and_devices(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -3373,12 +3387,16 @@ def test_resize_as_all_dtypes_and_devices(self):
     def test_view_all_dtypes_and_devices(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -3373,12 +3387,16 @@ def test_resize_as_all_dtypes_and_devices(self):
     def test_fill_all_dtypes_and_devices(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -3388,13 +3406,17 @@ def test_fill_all_dtypes_and_devices(self):
     def test_clone_all_dtypes_and_devices(self):
         for device in torch.testing.get_all_device_types():
             for dt in torch.testing.get_all_dtypes():
+                if (device == 'cuda' and dt == torch.bfloat16):
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -3409,6 +3431,11 @@ def test_tensor_factories_empty(self):
         for device in torch.testing.get_all_device_types():
             for shape in shapes:
                 for dt in torch.testing.get_all_dtypes():
+                    if dt == torch.bfloat16:
```
and here.

gchanan commented on test/test_torch.py:
```diff
@@ -10392,6 +10419,10 @@ def test_print(self):
+            if t == torch.cuda.BFloat16Tensor:
+                continue  # TODO: remove once bfloat16 is available on cuda
+            if t == torch.BFloat16Tensor:
+                continue  # TODO: Fix in the future PRs regarding bfloat16
```
wait why doesn't this work?

gchanan commented on c10/core/ScalarType.h:
```diff
@@ -77,7 +79,8 @@ namespace c10 {
   _(c10::quint8, QUInt8, i)       \
-  _(c10::qint32, QInt32, i)
+  _(c10::qint32, QInt32, i)       \
+  _(at::BFloat16, BFloat16, d)
```
wait, why doesn't this have bool?

izdeby replied:
AT_FORALL_SCALAR_TYPES(_) is not being used anywhere in our code base and i was actually going to remove BFloat16 from here as we have this macro for BC reasons. 
Should i add bool or should i remove BFloat16?

nairbv replied:
might be a question for @izdeby ?

gchanan replied:
if it's not being used anywhere, can't you just kill it?  (You can do this in a different PR if you want).

gchanan commented on c10/core/ScalarType.h:
```diff
@@ -87,7 +90,8 @@ namespace c10 {
   _(float, Float, d)                          \
-  _(double, Double, d)
+  _(double, Double, d)                        \
+  _(at::BFloat16, BFloat16, d)
```
same here.

gchanan replied:
could kill this one also?

izdeby replied:
actually i was wrong, it is used. Filed an [issue](https://github.com/pytorch/pytorch/issues/22668) for bool and assigned myself.

izdeby replied:
Closing the issue :) In Scalar.h there is overloaded boolean version. The overload needed because of issue with the SymbolicVariable. There is a comment: 

> "Value* is both implicitly convertible to SymbolicVariable and bool which causes ambiguosity error. Specialized constructor for bool resolves this problem."

gchanan commented on c10/util/BFloat16-inl.h:
```diff
@@ -15,3 +16,21 @@ inline C10_HOST_DEVICE BFloat16::operator float() const {
+
+namespace std {
+
+template <>
```
did you just put the minimal stuff here to get it to compile?  Because the actual struct has more, right?

gchanan replied:
this is fine, but you should probably file a task to fill this out completely.  It's reasonable for third parties to want to use this and it's surprising that all of the numeric_limits definitions aren't there.

izdeby replied:
[Filed](https://github.com/pytorch/pytorch/issues/22669).

gchanan commented on c10/util/BFloat16-inl.h:
```diff
@@ -15,3 +16,21 @@ inline C10_HOST_DEVICE BFloat16::operator float() const {
+    static constexpr bool has_infinity = true;
+    static constexpr bool has_quiet_NaN = true;
+    static constexpr c10::BFloat16 lowest() {
+      return at::BFloat16(0x0080, at::BFloat16::from_bits());
```
shouldn't this be FF7F?  The same as max but with the sign bit flipped.

izdeby commented on aten/src/ATen/Dispatch.h:
```diff
@@ -287,7 +298,25 @@ inline void deprecated_AT_DISPATCH_ALL_TYPES_AND_HALF_AND_COMPLEX() {}
+    }                                                                                                       \
+  }()
+
+#define AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(SCALARTYPE1, SCALARTYPE2, SCALARTYPE3, TYPE, NAME, ...)       \
```
Should i create a separate version for AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND to have 2 and 3 scalar types to be safe?

gchanan replied:
nah, I think it's fine.  I'd want to know if someone actually uses this so they can complain :).

izdeby committed:
Update on "Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

ezyang commented on aten/src/ATen/Dispatch.h:
```diff
@@ -36,6 +36,17 @@ struct ScalarTypeToCType<at::ScalarType::Half> {
+
+  // This is a workaround for the CUDA bug which prevents ::detail::ScalarTypeToCType<T>::type being used directly
+  // due to ambiguous reference which can't to be resolved. For some reason it cant pick between at::detail and at::cuda::detail.
+  // For repro example, please see: https://gist.github.com/izdeby/952ae7cf256ddb740a73776d39a7e7ba
```
Did we report this upstream to NVIDIA? Once you have an account, you can file a bug at https://developer.nvidia.com/nvidia_bug/add

izdeby replied:
I talked to @ngimel and she told me that she will file a bug which should have been fixed in cuda 10.2, but as we still have to support older versions of cuda, this hack is here to stay

ezyang commented on test/test_torch.py:
```diff
@@ -7984,17 +8017,17 @@ def test_scatter_bool(self):
             res = torch.zeros(3, 3, dtype=torch.bool, device=device)
             res = res.scatter_(0, torch.tensor([[0, 1, 2], [0, 1, 2]], device=device), x)
-            self.assertEqual(res, torch.tensor([[True, False, False], 
-                                                [False, True, False], 
```
Ideally, you'd fix the lint error in the earlier diffs.

izdeby committed:
Update on "Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

gchanan commented on test/common_utils.py:
```diff
@@ -507,7 +507,7 @@ def assertEqual(self, x, y, prec=None, message='', allow_inf=False):
                 super(TestCase, self).assertEqual(a.size(), b.size(), message)
                 if a.numel() > 0:
-                    if a.device.type == 'cpu' and a.dtype == torch.float16:
+                    if (a.device.type == 'cpu' and (a.dtype == torch.float16 or a.dtype == torch.bfloat16)):
```
the comment below is out of date.

gchanan commented on test/test_torch.py:
```diff
@@ -3389,13 +3411,21 @@ def test_clone_all_dtypes_and_devices(self):
             for dt in torch.testing.get_all_dtypes():
                 x = torch.tensor((1, 1), dtype=dt, device=device)
+                if (device == 'cuda' and dt == torch.bfloat16):
+                    self.assertRaises(RuntimeError, lambda: x.clone())
```
nit: for assertRaises, is there a message you can you set for when the tests fail?  It would be helpful to put something like "placeholder for future functionality -- check if you should update the test" so it's clear we don't actually intend for this to be the perfect behavior.

gchanan commented on test/test_torch.py:
```diff
@@ -3409,6 +3439,9 @@ def test_tensor_factories_empty(self):
                 for dt in torch.testing.get_all_dtypes():
+
+                    if (device == 'cuda' and dt == torch.bfloat16):
+                        continue  # Remove once cuda available for bfloat16
```
should there be an assertRaises here?

gchanan reviewed (approved):
just some follow on issues to file/fix.

izdeby committed:
Update on "Added Bfloat16 tensor for cpu with very limited support"

Added Bfloat16 tensor for cpu with very limited support

gh-metadata: pytorch pytorch 21860 gh/izdeby/10/head

ailzhang commented:
This PR broke lint.

izdeby commented:
> This PR broke lint.

@ailzhang, working of a fix PR
